### PR TITLE
[UASTC] Add random block generator

### PR DESCRIPTION
**Pull request by lexaknyazev**
This is a simple tool to generate as many valid UASTC blocks as one may need. It accepts UASTC mode index `[0..18]` and the number of blocks. This tool's memory layout is meant to be compatible with the transcoders, i.e. the same memory object could be used with both modules simultaneously.

**Standalone usage example, save blocks to disk**
```js
const fs = require('fs');

const totalBlocks = 8 * 1024 * 1024;
const memory = new WebAssembly.Memory({ initial: 2049 });
const view = new Uint8Array(memory.buffer, 65536);

const generatorModule = new WebAssembly.Module(fs.readFileSync('build/uastc_generator.wasm'));
const generatorInstance = new WebAssembly.Instance(generatorModule, { env: { memory: memory } });

for (let m = 0; m < 19; m++) {
    generatorInstance.exports.generate(m, totalBlocks);
    const filename = m.toString().padStart(2, '0') + '.uastc';
    fs.writeFileSync(filename, view);
}
```

**Combined usage example**
```js
const fs = require('fs');

const totalBlocks = 8 * 1024 * 1024;
const memory = new WebAssembly.Memory({ initial: 2049 });
const view = new Uint8Array(memory.buffer, 65536);

const generatorModule = new WebAssembly.Module(fs.readFileSync('build/uastc_generator.wasm'));
const generatorInstance = new WebAssembly.Instance(generatorModule, { env: { memory: memory } });

const transcoderModule = new WebAssembly.Module(fs.readFileSync('build/uastc_bc7.wasm'));
const transcoderInstance = new WebAssembly.Instance(transcoderModule, { env: { memory: memory } });

for (let m = 0; m < 19; m++) {
    generatorInstance.exports.generate(m, totalBlocks);

    // slice view and transcode it with the reference transcoder
    // ...

    // transcode with our transcoder in-place
    transcoderInstance.exports.transcode(totalBlocks);

    // compare results
    // ...
}
```

/cc @MaxGraey

### CODE REVIEW

**Review by donmccurdy — APPROVED** [review]

**Comment by donmccurdy** [comment]
Thanks, this looks great!

> // slice view and transcode it with the reference transcoder

For the purposes of unit tests, I'm not sure we should use the MSC transcoder in a CI environment yet. Or at least, I'd rather not attempt that until it's packaged and published to npm in a nodejs-compatible way. So far I've been manually opening a web browser and using the JS console to export sample output with it. I'm happy to use this transcoder as the source of truth for those tests, if we think its current state is correct and just want to prevent regressions.

**Comment by MaxGraey** [comment]
Btw instead use series of store for CRC table:
```ts
store<u64>(crcOffset, 0x0000000000000000, 0);
store<u64>(crcOffset, 0x7AD870C830358979, 8);
store<u64>(crcOffset, 0xF5B0E190606B12F2, 16);
...
```
you could store data in section without any extra overhead via `memory.data`:
```ts
const crcOffset = memory.data<u64>([
  0x0000000000000000, 
  0x7AD870C830358979,
  0xF5B0E190606B12F2,
  ...
]);
```

and use global basic memory offset for heap and static data via `memoryBase` flag: `asc -O3 ... --memoryBase=63488`

File: assembly/uastc-generator.ts [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+function crc64(crc: u64, seed: u64): u64 {
+  let count = 8;
+
+  while (count--) {
+    const byte = <u8>(crc ^ (seed & 0xFF));
+    crc = load<u64>(byte, crcOffset) ^ (crc >> 8);
+    seed >>= 8;
+  }
+  return crc;
+}
```
Comment (MaxGraey) [inline assembly/uastc-generator.ts:400]: Not sure it significant speedup but you could try unrolled version:
```suggestion
function crc64(crc: u64, seed: u64): u64 {
  let byte: usize;
  
  byte = usize((crc ^ (seed & 0xFF)) & 0xFF);
  crc = load<u64>(byte, crcOffset) ^ (crc >> 8);

  byte = usize((crc ^ (seed >> 8 & 0xFF)) & 0xFF);
  crc = load<u64>(byte, crcOffset) ^ (crc >> 8);

  byte = usize((crc ^ (seed >> 16 & 0xFF)) & 0xFF);
  crc = load<u64>(byte, crcOffset) ^ (crc >> 8);

  byte = usize((crc ^ (seed >> 24 & 0xFF)) & 0xFF);
  crc = load<u64>(byte, crcOffset) ^ (crc >> 8);

  byte = usize((crc ^ (seed >> 32 & 0xFF)) & 0xFF);
  crc = load<u64>(byte, crcOffset) ^ (crc >> 8);

  byte = usize((crc ^ (seed >> 40 & 0xFF)) & 0xFF);
  crc = load<u64>(byte, crcOffset) ^ (crc >> 8);

  byte = usize((crc ^ (seed >> 48 & 0xFF)) & 0xFF);
  crc = load<u64>(byte, crcOffset) ^ (crc >> 8);

  byte = usize((crc ^ (seed >> 58)) & 0xFF);
  crc = load<u64>(byte, crcOffset) ^ (crc >> 8);

  return crc;
}
```

Comment (lexaknyazev) [reply assembly/uastc-generator.ts:400]: The unrolled version seems to be ~8% faster.

Comment (MaxGraey) [reply assembly/uastc-generator.ts:400]: I was expected. So up to you apply this changes or leave all "as is" for simplicity

File: assembly/uastc-generator.ts [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+      case 18:
+        r0 = 0x09; //1001
+
+        r0 |= crc64(0, seed) << (4 + 15);
+        break;
+    }
+
+    const r1 = crc64(r0, seed);
+
+    const offset = 65536 + <u32>seed * 16;
```
Comment (MaxGraey) [inline assembly/uastc-generator.ts:124]:
```suggestion
    const offset = 2048 + <u32>seed * 16;
```
`2048` = `65536 - 63488`

Comment (lexaknyazev) [reply assembly/uastc-generator.ts:124]: Why? According to the docs, `--memoryBase` affects only memory segments (like static data or heap).

Comment (MaxGraey) [reply assembly/uastc-generator.ts:124]: You are right! Only for static data and managed heap)

Comment (MaxGraey) [reply assembly/uastc-generator.ts:124]: May be make sense do this like:
```ts
const offset = (ASC_MEMORY_BASE + 2048) + <u32>seed * 16;
```
WDYT?

Comment (lexaknyazev) [reply assembly/uastc-generator.ts:124]: Not sure about that as `65536` is used explicitly in all other modules of this project as the user data offset. Maybe we should provide it with `asc -u USER_DATA_OFFSET=65536` everywhere. 

As we reserve the zeroth memory page for internal needs, `63488` was chosen to put generator's data to the very end of it so the generator and the transcoders could use the same memory without risking any overlap (although currently implemented UASTC modules use only up to 6kB of 64kB).

**Pull request merged**